db/src/test/db.spec.ts

```diff
@@ -73,24 +76,161 @@ test('DB.remove removes existing key from DB and returns true', async (t) => {
 });
 
-test('DB.update updates an existing document', async (t) => {
+test('DB.update updates an existing document, returns it, and increments version', async (t) => {
   const { db } = t.context;
   await db.create('test', { b: 2 });
   const next = await db.update('test', (prev) => ({ ...prev, a: 1 }));
-  const val = await db.get('test');
-  t.deepEqual(val, { a: 1, b: 2 });
-  t.deepEqual(next, val);
+  const { value, version } = (await db.getWithMeta('test'))!;
+  t.deepEqual(value, { a: 1, b: 2 });
+  t.deepEqual(next, value);
+  t.is(version, 2);
+});
+
+test('DB.update does nothing if document not updated', async (t) => {
+  const { db } = t.context;
+  await db.create('test', { b: 2 });
+  const next = await db.update('test', (prev) => ({ ...prev }));
+  const { value, version } = (await db.getWithMeta('test'))!;
+  t.deepEqual(value, { b: 2 });
+  t.deepEqual(next, value);
+  t.is(version, 1);
+});
+
+test('DB.poll returns patches which match requested versions', async (t) => {
+  const { db } = t.context;
+  await db.create('test1', 'a');
+  await db.create('test2', 'a');
+  await db.create('test3', 'a');
+  await db.create('test4', 'a');
+  await db.update('test1', () => 'b');
+  await db.update('test1', () => 'c');
+  await db.update('test2', () => 'b');
+  await db.update('test2', () => 'c');
+  await db.update('test3', () => 'b');
+  await db.remove('test3');
+  const keyedPatches = await db.poll([
+    ['test1', 1], ['test2', 2], ['test3', 1], ['test4', 1],
+  ]);
+  t.deepEqual(keyedPatches, [
+    ['test1', [
+      { version: 2, ops: [{ op: 'replace', path: '/root', value: 'b' }] },
+      { version: 3, ops: [{ op: 'replace', path: '/root', value: 'c' }] },
+    ]],
+    ['test2', [
+      { version: 3, ops: [{ op: 'replace', path: '/root', value: 'c' }] },
+    ]],
+    ['test3', [
+      { version: 2, ops: [{ op: 'replace', path: '/root', value: 'b' }] },
+      { version: 3, ops: [{ op: 'remove', path: '/root' }] },
+    ]],
+  ]);
+});
+
+test('DB.poll returns on update if no new patches stored', async (t) => {
+  const { db } = t.context;
+  await db.create('test1', 'a');
+  const [keyedPatches] = await Promise.all([
+    db.poll([['test1', 1]]),
+    db.update('test1', () => 'b'),
+  ]);
+  t.deepEqual(keyedPatches, [
+    ['test1', [
+      { version: 2, ops: [{ op: 'replace', path: '/root', value: 'b' }] },
+    ]],
+  ]);
+});
+
+test('DB.poll returns on remove if no new patches stored', async (t) => {
+  const { db } = t.context;
+  await db.create('test1', 'a');
+  const [keyedPatches] = await Promise.all([
+    db.poll([['test1', 1]]),
+    db.remove('test1'),
+  ]);
+  t.deepEqual(keyedPatches, [
+    ['test1', [
+      { version: 2, ops: [{ op: 'remove', path: '/root' }] },
+    ]],
+  ]);
+});
+
+test('DB.create works after remove', async (t) => {
+  const { db } = t.context;
+  const t0 = Date.now();
+  await db.create('test', 7);
+  await db.remove('test');
+  t.true(await db.create('test', 8));
```

Comment:
This test may fail on some eventually consistent datastores: it's perfectly reasonable for the second `create` to fail because the key still exists.

Reply:
Not on the local DB though..

Reply:
OK.